File: dodam-core/src/main/java/b1nd/dodamcore/outgoing/application/OutGoingService.java
package b1nd.dodamcore.outgoing.application;

import b1nd.dodamcore.member.domain.entity.Student;
import b1nd.dodamcore.outgoing.domain.entity.OutGoing;
import b1nd.dodamcore.outgoing.domain.exception.OutGoingNotFoundException;
import b1nd.dodamcore.outgoing.repository.OutGoingRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;

@Service
@RequiredArgsConstructor
public class OutGoingService {

    private final OutGoingRepository repository;

    public void save(OutGoing outGoing) {
        repository.save(outGoing);
    }

    public void delete(OutGoing outGoing) {
        repository.delete(outGoing);
    }

    public OutGoing getById(Long id) {
        return repository.findById(id)
                .orElseThrow(OutGoingNotFoundException::new);
    }

    public List<OutGoing> getByDate(LocalDateTime startDay, LocalDateTime endDay) {
        return repository.findByDate(startDay, endDay);
    }

    public List<OutGoing> getByStudent(Student student, LocalDateTime now) {
        return repository.findByStudentAndEndAtGreaterThanEqual(student, now);
    }

}

File: services/service-out-going/src/main/kotlin/com/b1nd/dodamdodam/outgoing/domain/outgoing/service/OutGoingService.kt
package com.b1nd.dodamdodam.outgoing.domain.outgoing.service

import com.b1nd.dodamdodam.outgoing.domain.outgoing.entity.OutGoingEntity
import com.b1nd.dodamdodam.outgoing.domain.outgoing.enumeration.OutGoingStatus
import com.b1nd.dodamdodam.outgoing.domain.outgoing.exception.OutGoingAlreadyExistsException
import com.b1nd.dodamdodam.outgoing.domain.outgoing.exception.OutGoingForbiddenException
import com.b1nd.dodamdodam.outgoing.domain.outgoing.exception.OutGoingNotFoundException
import com.b1nd.dodamdodam.outgoing.domain.outgoing.exception.OutGoingNotPendingException
import com.b1nd.dodamdodam.outgoing.domain.outgoing.repository.OutGoingRepository
import org.springframework.stereotype.Service
import java.time.LocalDate
import java.util.UUID

@Service
class OutGoingService(
    private val repository: OutGoingRepository
) {
    fun create(entity: OutGoingEntity): OutGoingEntity {
        if (repository.existsOverlapping(entity.studentId, entity.startAt, entity.endAt)) {
            throw OutGoingAlreadyExistsException()
        }
        return repository.save(entity)
    }

    fun findById(id: Long): OutGoingEntity =
        repository.findById(id).orElseThrow { OutGoingNotFoundException() }

    fun allow(id: Long) {
        val entity = findById(id)
        entity.status = OutGoingStatus.ALLOWED
    }

    fun reject(id: Long, rejectReason: String?) {
        val entity = findById(id)
        entity.status = OutGoingStatus.REJECTED
        entity.rejectReason = rejectReason
    }

    fun revert(id: Long) {
        val entity = findById(id)
        entity.status = OutGoingStatus.PENDING
        entity.rejectReason = null
    }

    fun delete(id: Long, studentId: UUID) {
        val entity = findById(id)
        if (entity.studentId != studentId) throw OutGoingForbiddenException()
        if (entity.status != OutGoingStatus.PENDING) throw OutGoingNotPendingException()
        repository.delete(entity)
    }

    fun findByDate(date: LocalDate): List<OutGoingEntity> =
        repository.findByStartAtLessThanEqualAndEndAtGreaterThanEqual(
            date.atTime(23, 59, 59),
            date.atStartOfDay()
        )

    fun findByStudentId(studentId: UUID): List<OutGoingEntity> =
        repository.findByStudentId(studentId)
}
